src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ForeachToFor
+{
+    internal abstract class AbstractForeachToForCodeRefactoringProvider
+        : CodeRefactoringProvider
+    {
+        private const string get_Count = nameof(get_Count);
+        private const string get_Item = nameof(get_Item);
+
+        private const string Length = nameof(Length);
+        private const string Count = nameof(Count);
+
+        private static readonly ImmutableArray<string> s_KnownInterfaceNames =
+            new string[] { typeof(IList<>).FullName, typeof(IReadOnlyList<>).FullName, typeof(IList).FullName }.ToImmutableArray();
```

Comment:
we should also support foreaching over strings .

Reply:
sure.

Reply:
just an idea: what if this was pattern-based? meaning we would look for a either Length or Count property + an indexer as opposed to a fixed set of interfaces

Reply:
people could of course implement this properties however they want so the result could technically be wrong, but they can already do that when implementing these interfaces... although it's true that at least they're explicitly saying what the purpose of a Count property is by implementing the interface so I'm not entirely sure if it's a good idea.. @CyrusNajmabadi what are your thoughts on this?

Reply:
Yes.  As a refactoring, i think this should be lenient and should be pattern-based.  That's what i'm doing for the for-to-foreach refactoring.

Reply:
added to issue - https://github.com/dotnet/roslyn/issues/25667